drivers/afe/ad4110/ad4110.c

```diff
@@ -530,66 +705,156 @@ int32_t ad4110_setup(struct ad4110_dev **device,
-		ret |= ad4110_spi_int_reg_write_msk(dev,
-						    A4110_ADC,
-						    AD4110_REG_ADC_INTERFACE,
-						    AD4110_DATA_STAT_EN,
-						    AD4110_REG_ADC_INTERFACE_DS_MSK);
-	}
-	dev->data_stat = init_param.data_stat;
-
-	if(init_param.data_length == AD4110_DATA_WL16) {
-		ret |= ad4110_spi_int_reg_write_msk(dev,
-						    A4110_ADC,
-						    AD4110_REG_ADC_INTERFACE,
-						    AD4110_DATA_WL16,
-						    AD4110_REG_ADC_INTERFACE_WL16_MSK);
-	}
-	dev->data_length = init_param.data_length;
-
 	if(init_param.afe_crc_en != AD4110_AFE_CRC_DISABLE) {
-		ret |= ad4110_spi_int_reg_write(dev,
-						A4110_AFE,
-						AD4110_REG_AFE_CNTRL1,
-						AD4110_REG_AFE_CNTRL1_CRC_EN);
+		ret = ad4110_spi_int_reg_write(dev,
+					       A4110_AFE,
+					       AD4110_REG_AFE_CNTRL1,
+					       AD4110_REG_AFE_CNTRL1_CRC_EN);
+		if (ret)
+			return ret;
 	}
 	dev->afe_crc_en = init_param.afe_crc_en;
 
 	if(init_param.adc_crc_en != AD4110_ADC_CRC_DISABLE) {
-		ret |= ad4110_spi_int_reg_write_msk(dev,
-						    A4110_ADC,
-						    AD4110_REG_ADC_INTERFACE,
-						    AD4110_ADC_CRC_EN(init_param.adc_crc_en),
-						    AD4110_REG_ADC_INTERFACE_CRC_EN_MSK);
+		ret = ad4110_spi_int_reg_write_msk(dev,
+						   A4110_ADC,
+						   AD4110_REG_ADC_INTERFACE,
+						   AD4110_ADC_CRC_EN(init_param.adc_crc_en),
+						   AD4110_REG_ADC_INTERFACE_CRC_EN_MSK);
+		if (ret)
+			return ret;
 	}
 	dev->adc_crc_en = init_param.adc_crc_en;
 
-	ret |= ad4110_set_op_mode(dev, init_param.op_mode);
-	dev->op_mode = init_param.op_mode;
+	if(dev->data_stat == AD4110_ENABLE) {
+		ret = ad4110_spi_int_reg_write_msk(dev,
+						   A4110_ADC,
+						   AD4110_REG_ADC_INTERFACE,
+						   AD4110_DATA_STAT_EN,
+						   AD4110_REG_ADC_INTERFACE_DS_MSK);
+		if (ret)
+			return ret;
+	}
+
+	if(dev->data_length == AD4110_DATA_WL16) {
+		ret = ad4110_spi_int_reg_write_msk(dev,
+						   A4110_ADC,
+						   AD4110_REG_ADC_INTERFACE,
+						   AD4110_DATA_WL16,
+						   AD4110_REG_ADC_INTERFACE_WL16_MSK);
+		if (ret)
+			return ret;
+	}
+
+	if(dev->sync != AD4110_SYNC_EN) {
+		ret = ad4110_spi_int_reg_write_msk(dev,
+						   A4110_ADC,
+						   AD4110_REG_ADC_GPIO_CONFIG,
+						   AD4110_REG_GPIO_CONFIG_SYNC_EN(AD4110_SYNC_DIS),
+						   AD4110_REG_GPIO_CONFIG_SYNC_EN(0xF));
+		if (ret)
+			return ret;
+	}
 
-	ret |= ad4110_set_gain(dev, init_param.gain);
-	dev->gain = init_param.gain;
+
+	ret = ad4110_set_op_mode(dev, dev->op_mode);
+	if (ret)
+		return ret;
+
+	ret = ad4110_set_gain(dev, dev->gain);
+	if (ret)
+		return ret;
+
+	ret = ad4110_set_reference(dev, dev->volt_ref);
+	if (ret)
+		return ret;
+
+	/* When AD4110_AFE_ADC_CLOCKED selected, adc_clk must be AD4110_ADC_INT_CLK_CLKIO */
+	if((dev->afe_clk == AD4110_AFE_ADC_CLOCKED)
+	    && (dev->adc_clk != AD4110_ADC_INT_CLK_CLKIO))
+		return FAILURE;
+	else {
+		ret = ad4110_set_adc_clk(dev, dev->adc_clk);
+		if (ret)
+			return ret;
+	}
+
+	ret = ad4110_set_afe_clk(dev, dev->afe_clk);
```

Comment:
I guess you can return here directly if error code is encountered.
`*device = dev;` will be redundant.

Maybe move here the following if/else statement.
```
if (!ret) {
		printf("AD4110 successfully initialized\n");
} else {
		pr_err("AD4110 initialization error (%d)\n", ret);
                return ret;
}

*device = dev;

return SUCCESS;
```

Reply:
Actually, you might need to do a conditional jump specific for the init functions of each driver where you want to `free(dev)`. So probably all `return ret;` from these functions will be replaced by something like `goto err_init;`